Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp

```diff
@@ -348,40 +347,45 @@ static const char* fragmentTemplateCommon =
 
         void applyTextureCopy(inout vec4 color, vec2 texCoord)
         {
-            vec2 coord = clamp(texCoord, u_texelSize / 2., vec2(1., 1.) - u_texelSize / 2.);
-            coord = (u_textureCopyMatrix * vec4(coord, 0., 1.)).xy;
-            color = texture2D(s_sampler, coord);
-        }
+            vec2 min = (u_textureSpaceMatrix * vec4(0., 0., 0., 1.)).xy + u_texelSize / 2.;
+            vec2 max = (u_textureSpaceMatrix * vec4(1., 1., 0., 1.)).xy - u_texelSize / 2.;
 
-        vec4 sampleTextureClamp(sampler2D sampler, vec2 texCoord, vec2 offset, mat4 matrix)
-        {
-            vec2 coord = texCoord + offset;
-            coord = clamp(coord, u_texelSize / 2., vec2(1., 1.) - u_texelSize / 2.);
-            coord = (matrix * vec4(coord, 0., 1.)).xy;
-            return texture2D(sampler, coord);
+            vec2 coord = clamp(texCoord, min, max);
+
+            color = texture2D(s_sampler, coord);
         }
 
         void applyBlurFilter(inout vec4 color, vec2 texCoord)
         {
-            vec4 total = sampleTextureClamp(s_sampler, texCoord, vec2(0., 0.), u_textureBlurMatrix) * u_gaussianKernel[0];
+            vec2 step = u_blurDirection * u_texelSize;
+            vec2 min = (u_textureBlurMatrix * vec4(0., 0., 0., 1.)).xy + u_texelSize / 2.;
+            vec2 max = (u_textureBlurMatrix * vec4(1., 1., 0., 1.)).xy - u_texelSize / 2.;
```

Comment:
I belive most graphics cards optimize this part so that the calculations happen only once over all fragments.